### PR TITLE
Fix to save the state of FTRL models

### PR DESCRIPTION
**Pull request by bremen79**
I have made a fix to save the state of FTRL models. Currently the state is not saved at all because it calls the the gd save function that doesn't save the state unless it finds a gd structure, that is absent in ftrl models.

I have added a integer in the vw structure the indicates the size of the ftrl model to save and changed read/write of gd accordingly.

Note that the saving for Pistol and Coin betting still does not work: It correctly save the state, but the test fails. Not sure why, I think it is the loss in precision due to the saving and the exponential nature of these algorithms. On the other hand, the saving state for FTRL Proximal now works perfectly.

### CODE REVIEW

File: test/save_resume_test.py [hunk 1]
```diff
@@ -153,8 +153,9 @@ def do_test(filename, args, verbose=None, repeat_args=None, known_failure=False)
         errors += do_test(filename, '--loss_function logistic --link logistic')
         errors += do_test(filename, '--nn 2')
         errors += do_test(filename, '--binary')
-        errors += do_test(filename, '--ftrl', known_failure=True)
+        errors += do_test(filename, '--ftrl')
         errors += do_test(filename, '--pistol', known_failure=True)
```
Comment (JohnLangford) [inline test/save_resume_test.py]: Are pistol and coin still in known failure?

File: vowpalwabbit/global_data.h [hunk 2]
```diff
@@ -565,6 +565,7 @@ struct vw
   bool adaptive;            // Should I use adaptive individual learning rates?
   bool normalized_updates;  // Should every feature be normalized
   bool invariant_updates;   // Should we use importance aware/safe updates
+  uint32_t ftrl_size;
```
Comment (JohnLangford) [inline vowpalwabbit/global_data.h]: Could we pass this as an argument to save_load instead?  That seems more elegant than sticking it in the global data structure.

**Comment by JohnLangford** [comment]
The build 67 failure is fixed in the current patch that I'm working on.

File: vowpalwabbit/global_data.h (hunk 2 above)
Comment (bremen79) [reply vowpalwabbit/global_data.h]: Done.

File: test/save_resume_test.py (hunk 1 above)
Comment (bremen79) [reply test/save_resume_test.py]: Yes. As I write above, the data structure is saved correctly, but the weights seem to be different if trained continuously or trained, saved, resumed. The difference does not happen immediately, but after some samples. This is why I suspect numerical issues.

Comment (JohnLangford) [reply test/save_resume_test.py]: This violates my understanding of computers so I suspect there is something that we're missing.  I probably won't be able to debug before the next release, but it should get done...

Comment (bremen79) [reply test/save_resume_test.py]: I am trying to debug this issue a bit more

File: vowpalwabbit/global_data.h (hunk 2 above)
Comment (JohnLangford) [reply vowpalwabbit/global_data.h]: Oops, I merged.   Should I revert, or do you want to do a separate pull request?

Comment (bremen79) [reply vowpalwabbit/global_data.h]: Found the bug, not sure what best way to fix it, see mail

**Pull request merged**